pkg/services/ngalert/tests/fakes/rules.go

```diff
@@ -0,0 +1,341 @@
+package fakes
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"sync"
+	"testing"
+	"time"
+
+	models2 "github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/grafana/grafana/pkg/services/user"
+	"github.com/grafana/grafana/pkg/util"
+)
+
+// FakeRuleStore mocks the RuleStore of the scheduler.
+type FakeRuleStore struct {
+	t   *testing.T
+	mtx sync.Mutex
+	// OrgID -> RuleGroup -> Namespace -> Rules
+	Rules       map[int64][]*models.AlertRule
+	Hook        func(cmd interface{}) error // use Hook if you need to intercept some query and return an error
+	RecordedOps []interface{}
```

Comment:
I grepped through, and we only verify RecordedOps in a couple of places. I don't think it's a good system - It would be better to just verify that the state of the database is as expected than cart around this unwieldy RecordedOps setup, switching on the arg type and counting against expectations. What you should see is that a rule you wanted is no longer available, not the # of times you called `Delete`.

Let's file a ticket to nix this thing.

Reply:
Recorded Ops, as well as Hook below, are the sign that we need just a mock instead of fake. 

> to just verify that the state of the database is as expected than cart around this unwieldy RecordedOps setup,

Then to verify a state the fake would have to simulate the database. This is even more code in the fake and logic that also needs to be tested (tests for fakes). I would rather replace it with mocks.

Reply:
My first instinct with this was actually to just go ahead and refactor tests to use a mock rather than move the store.

I actually got semi-far with this. A lot of tests pass with the mock. Some edge cases were not so easy, though. Some of our tests rely heavily on the filtering logic in the fake (which means they're testing the fake itself, not the prod code lol). You'd `PutRule`, and then `ListRules` would only respond with a subset of the expected rules because of the filtering.

These were pretty difficult to pull apart, so I put that branch on hold for now.

Just noting that I agree with the two of you - this was the thing I tried first.